engine/engine_impl.go

```diff
@@ -246,6 +252,12 @@ func (k *Kubernetes) tail(ctx context.Context, spec *Spec, step *Step, output io
 
 func (k *Kubernetes) start(spec *Spec, step *Step) error {
 	err := retry.RetryOnConflict(backoff, func() error {
+		// We protect this read/modify/write with a mutex to reduce the
+		// chance of a self-inflicted concurrent modification error
+		// when a DAG in a pipeline is fanning out and we have a lot of
+		// steps to Start at once.
+		k.updateMutex.Lock()
+		defer k.updateMutex.Unlock()
```

Comment:
when we introduce mutexes in drone we tend to be overly cautious in doing so. Specifically we should make sure we can recover from a panic and unlock the mutex. Example here: https://github.com/drone/runner-go/blob/df14cddb19a2f9398473bfdc78e7294e227afb2a/client/single.go#L39:L50

Reply:
My understanding is that defers execute as the panic unwinds the stack. The recover call itself is within a defer...

Reply:
On the other hand...we're inside a naked goroutine inside `execer.exec` (from natessilva/dag), so we should have a top-level `recover()` if there can be panics...otherwise the entire process is going to crash. Same with the goroutine we spawn around line 179 for the detached use case.

Reply:
(Sorry, that previous comment was referring to `runtime/execer.go`...I was looking at existing semaphore usage.)

Reply:
I believe a recover is required so that pipeline execution can gracefully fail and send updates to the Drone server.

Reply:
Sorry...I don't understand the ask. Do we want to start swallowing any panics thrown in `(*Kubernetes).start`, where we didn't swallow them before? Is that the right level at which to swallow them? What's different about `start` compared to `tail`, `waitForReady`, etc., such that we should swallow panics in it?

Reply:
you are correct.  defer always executes even on panic and we and this pull request does not introduce any new risk that did not exist before.

I would like to audit the code and make sure we gracefully recover from panics, however, that is no reason to block this particular pull request.

Reply:
Totally agree. Thank you 👍